Transient "disposable" service language updates (#31738)
dotnet/AspNetCore.Docs: aspnetcore/blazor/fundamentals/dependency-injection.md

## Utility base component classes to manage a DI scope

<!-- UPDATE 10.0 - PU design is under consideration for .NET 10. -->

In non-Blazor ASP.NET Core apps, scoped and transient services are typically scoped to the current request. After the request completes, scoped and transient services are disposed by the DI system.

> In interactive server-side Blazor apps, the request scope lasts for the duration of the circuit (the SignalR connection between the client and server), which can result in scoped and disposable transient services living much longer than the lifetime of a single component. Therefore, don't directly inject a scoped service into a component if you intend the service lifetime to match the lifetime of the component. Transient services injected into a component that don't implement <xref:System.IDisposable> are garbage collected when the component is disposed. However, injected transient services *that implement <xref:System.IDisposable>* are maintained by the DI container for the lifetime of the circuit, which prevents service garbage collection when the component is disposed and results in a memory leak. An alternative approach for scoped and disposable transient services based on the <xref:Microsoft.AspNetCore.Components.OwningComponentBase> type is described later in this section. For more information, see [Design for solving transient disposables on Blazor Server (`dotnet/aspnetcore` #26676)](https://github.com/dotnet/aspnetcore/issues/26676).

Author: ```suggestion
In interactive server-side Blazor apps, the request scope lasts for the duration of the circuit (the SignalR connection between the client and server), which can result in scoped and disposable transient services living much longer than the lifetime of a single component. Therefore, don't directly inject a scoped service into a component if you intend the service lifetime to match the lifetime of the component. Transient services injected into a component that don't implement <xref:System.IDisposable> are garbage collected when the component is disposed. However, injected transient services *that implement <xref:System.IDisposable>* are maintained by the DI container for the lifetime of the circuit, which prevents service garbage collection when the component is disposed and results in a memory leak. An alternative approach for scoped services based on the <xref:Microsoft.AspNetCore.Components.OwningComponentBase> type is described later in this section, and disposable transient services shouldn't be used at all. For more information, see [Design for solving transient disposables on Blazor Server (`dotnet/aspnetcore` #26676)](https://github.com/dotnet/aspnetcore/issues/26676).
```

> Even in client-side Blazor apps that don't operate over a circuit, services registered with a scoped lifetime are treated as singletons, so they live longer than scoped services in typical ASP.NET Core apps. Client-side disposable transient services can also live longer than expected because the DI container persists for the lifetime of the app. Although long-lived disposable transient services are of greater concern on the server, they should be avoided as client service registrations as well. Use of the <xref:Microsoft.AspNetCore.Components.OwningComponentBase> type is also recommended for client-side services to control service lifetime.

Reviewer: >Client-side disposable transient services can also live longer than expected because the DI container persists for the lifetime of the app.

=>

>Client-side disposable transient services can also live longer than expected because the DI container, which holds references to disposable services, persists for the lifetime of the app, preventing them from being garbage-colleced.

...but double-check with @MackinnonBuck please.

> Even in client-side Blazor apps that don't operate over a circuit, services registered with a scoped lifetime are treated as singletons, so they live longer than scoped services in typical ASP.NET Core apps. Client-side disposable transient services can also live longer than expected because the DI container persists for the lifetime of the app. Although long-lived disposable transient services are of greater concern on the server, they should be avoided as client service registrations as well. Use of the <xref:Microsoft.AspNetCore.Components.OwningComponentBase> type is also recommended for client-side services to control service lifetime.

Author: ```suggestion
Even in client-side Blazor apps that don't operate over a circuit, services registered with a scoped lifetime are treated as singletons, so they live longer than scoped services in typical ASP.NET Core apps. Client-side disposable transient services also live longer than the components where they're injected because the DI container, which holds references to disposable services, persists for the lifetime of the app, preventing garbage collection on the services. Although long-lived disposable transient services are of greater concern on the server, they should be avoided as client service registrations as well. Use of the <xref:Microsoft.AspNetCore.Components.OwningComponentBase> type is also recommended for client-side services to control service lifetime.
```

> Even in client-side Blazor apps that don't operate over a circuit, services registered with a scoped lifetime are treated as singletons, so they live longer than scoped services in typical ASP.NET Core apps. Client-side disposable transient services also live longer than the components where they're injected because the DI container, which holds references to disposable services, persists for the lifetime of the app, preventing garbage collection on the services. Although long-lived disposable transient services are of greater concern on the server, they should be avoided as client service registrations as well. Use of the <xref:Microsoft.AspNetCore.Components.OwningComponentBase> type is also recommended for client-side services to control service lifetime.

Author: ```suggestion
Even in client-side Blazor apps that don't operate over a circuit, services registered with a scoped lifetime are treated as singletons, so they live longer than scoped services in typical ASP.NET Core apps. Client-side disposable transient services also live longer than the components where they're injected because the DI container, which holds references to disposable services, persists for the lifetime of the app, preventing garbage collection on the services. Although long-lived disposable transient services are of greater concern on the server, they should be avoided as client service registrations as well. Use of the <xref:Microsoft.AspNetCore.Components.OwningComponentBase> type is also recommended for client-side scoped services to control service lifetime, and disposable transient services shouldn't be used at all.
```

> An approach that limits a service lifetime is use of the <xref:Microsoft.AspNetCore.Components.OwningComponentBase> type. <xref:Microsoft.AspNetCore.Components.OwningComponentBase> is an abstract type derived from <xref:Microsoft.AspNetCore.Components.ComponentBase> that creates a DI scope corresponding to the *lifetime of the component*. Using this scope instead of a scoped service or disposable transient service, a component can inject services with a scoped lifetime and have them live as long as the component. When the component is destroyed, services from the component's scoped service provider are disposed as well. This can be useful for services that:

Reviewer: > Using this scope instead of a scoped service or disposable transient service, a component can inject services with a scoped lifetime and have them live as long as the component.

=>

> Using this scope for a scoped service or disposable transient service, a component can inject services with a limited lifetime and have them live as long as the component.

Author: I think I like mine better. 😄 I don't want to imply that `ScopedService` should try to take a disposable transient service. Therefore, I like "instead of" ... ***however*** I do want to change it to use `OwningComponentBase` over "scope" ...

```suggestion
An approach that limits a service lifetime is use of the <xref:Microsoft.AspNetCore.Components.OwningComponentBase> type. <xref:Microsoft.AspNetCore.Components.OwningComponentBase> is an abstract type derived from <xref:Microsoft.AspNetCore.Components.ComponentBase> that creates a DI scope corresponding to the *lifetime of the component*. Using this scope instead of directly injecting a scoped service or using a disposable transient service, a component can inject services with a scoped lifetime and have them live as long as the component. When the component is destroyed, services from the component's scoped service provider are disposed as well. This can be useful for services that:
```

Author: It can't AFAIK ... AFAIK transient disposables shouldn't be used at all, and that's what the text says. However, @MackinnonBuck can set me straight on it.

Reviewer: I did the test and `OwningComponentBase` disposes the transient disposable.
- [TransientDisposableService.cs](https://github.com/hakenr/BlazorTransientDisposablesTest/blob/master/BlazorTransientDisposables/BlazorTransientDisposables.Client/TransientDisposableService.cs)
- [Home.razor](https://github.com/hakenr/BlazorTransientDisposablesTest/blob/master/BlazorTransientDisposables/BlazorTransientDisposables.Client/Pages/Home.razor)
- [Program.cs](https://github.com/hakenr/BlazorTransientDisposablesTest/blob/master/BlazorTransientDisposables/BlazorTransientDisposables.Client/Program.cs)

![image](https://github.com/dotnet/AspNetCore.Docs/assets/12828888/c2fa9bec-cf06-446e-8b2a-c18ddd9c5d15)

So I recommend to express you can use it to "fix" the evil transient disposable (coming from code out of your control).
;-)

Author: I see .....

Mackinnon might be able to make a ruling on this ... Javier originally said ***Do NOT use them on PAIN of DEATH 💀!*** .... well 😆 ... I added the "pain of death" part 😆. If you're not sure Mackinnon, I have an offline email convo with Javier from earlier that I can ping him on to ask a follow-up question about this ... can `OwningComponentBase`/`ScopedServices` be safely used for disposable transients?

> An approach that limits a service lifetime is use of the <xref:Microsoft.AspNetCore.Components.OwningComponentBase> type. <xref:Microsoft.AspNetCore.Components.OwningComponentBase> is an abstract type derived from <xref:Microsoft.AspNetCore.Components.ComponentBase> that creates a DI scope corresponding to the *lifetime of the component*. Using this scope instead of directly injecting a scoped service or using a disposable transient service, a component can inject services with a scoped lifetime and have them live as long as the component. When the component is destroyed, services from the component's scoped service provider are disposed as well. This can be useful for services that:

Author: ```suggestion
An approach that limits a service lifetime is use of the <xref:Microsoft.AspNetCore.Components.OwningComponentBase> type. <xref:Microsoft.AspNetCore.Components.OwningComponentBase> is an abstract type derived from <xref:Microsoft.AspNetCore.Components.ComponentBase> that creates a DI scope corresponding to the *lifetime of the component*. Using this scope for a scoped service, a component can inject services with a scoped lifetime and have them live as long as the component. When the component is destroyed, services from the component's scoped service provider are disposed as well. This can be useful for services that:
```

## Utility base component classes to manage a DI scope

<!-- UPDATE 10.0 - PU design is under consideration for .NET 10. -->

In non-Blazor ASP.NET Core apps, scoped and transient services are typically scoped to the current request. After the request completes, scoped and transient services are disposed by the DI system.

> In interactive server-side Blazor apps, the request scope lasts for the duration of the circuit (the SignalR connection between the client and server), which can result in scoped and disposable transient services living much longer than the lifetime of a single component. Therefore, don't directly inject a scoped service into a component if you intend the service lifetime to match the lifetime of the component. Transient services injected into a component that don't implement <xref:System.IDisposable> are garbage collected when the component is disposed. However, injected transient services *that implement <xref:System.IDisposable>* are maintained by the DI container for the lifetime of the circuit, which prevents service garbage collection when the component is disposed and results in a memory leak. An alternative approach for scoped services based on the <xref:Microsoft.AspNetCore.Components.OwningComponentBase> type is described later in this section, and disposable transient services shouldn't be used at all. For more information, see [Design for solving transient disposables on Blazor Server (`dotnet/aspnetcore` #26676)](https://github.com/dotnet/aspnetcore/issues/26676).

Reviewer 2: I think saying "DI scope" rather than "request scope" is more accurate, because there's no "request" that the DI scope is bound to.
```suggestion
In interactive server-side Blazor apps, the DI scope lasts for the duration of the circuit (the SignalR connection between the client and server), which can result in scoped and disposable transient services living much longer than the lifetime of a single component. Therefore, don't directly inject a scoped service into a component if you intend the service lifetime to match the lifetime of the component. Transient services injected into a component that don't implement <xref:System.IDisposable> are garbage collected when the component is disposed. However, injected transient services *that implement <xref:System.IDisposable>* are maintained by the DI container for the lifetime of the circuit, which prevents service garbage collection when the component is disposed and results in a memory leak. An alternative approach for scoped services based on the <xref:Microsoft.AspNetCore.Components.OwningComponentBase> type is described later in this section, and disposable transient services shouldn't be used at all. For more information, see [Design for solving transient disposables on Blazor Server (`dotnet/aspnetcore` #26676)](https://github.com/dotnet/aspnetcore/issues/26676).
```

Even in client-side Blazor apps that don't operate over a circuit, services registered with a scoped lifetime are treated as singletons, so they live longer than scoped services in typical ASP.NET Core apps. Client-side disposable transient services also live longer than the components where they're injected because the DI container, which holds references to disposable services, persists for the lifetime of the app, preventing garbage collection on the services. Although long-lived disposable transient services are of greater concern on the server, they should be avoided as client service registrations as well. Use of the <xref:Microsoft.AspNetCore.Components.OwningComponentBase> type is also recommended for client-side scoped services to control service lifetime, and disposable transient services shouldn't be used at all.

> An approach that limits a service lifetime is use of the <xref:Microsoft.AspNetCore.Components.OwningComponentBase> type. <xref:Microsoft.AspNetCore.Components.OwningComponentBase> is an abstract type derived from <xref:Microsoft.AspNetCore.Components.ComponentBase> that creates a DI scope corresponding to the *lifetime of the component*. Using this scope for a scoped service, a component can inject services with a scoped lifetime and have them live as long as the component. When the component is destroyed, services from the component's scoped service provider are disposed as well. This can be useful for services that:

Reviewer 2: > Using this scope for a scoped service, a component can inject services with a scoped lifetime and have them live as long as the component.

Hm, how about this?

_Using this scope, a component can inject services with a scoped lifetime and have them live as long as the component_

Author: ```suggestion
An approach that limits a service lifetime is use of the <xref:Microsoft.AspNetCore.Components.OwningComponentBase> type. <xref:Microsoft.AspNetCore.Components.OwningComponentBase> is an abstract type derived from <xref:Microsoft.AspNetCore.Components.ComponentBase> that creates a DI scope corresponding to the *lifetime of the component*. Using this scope, a component can inject services with a scoped lifetime and have them live as long as the component. When the component is destroyed, services from the component's scoped service provider are disposed as well. This can be useful for services that:
```

> * Should be reused within a component, as the transient lifetime is inappropriate with disposable transient services.

Reviewer 2: Was the "with disposable transient services" bit added to clarify that only _disposable_ transient services are inappropriate because they live longer than the lifetime of the component? I wonder if we could simplify the guidance by combining these two bullets to say:

_This can be useful for services that should be reused within a component, but shouldn't be shared across components._

> * Shouldn't be shared across components, as the singleton lifetime is inappropriate.

Reviewer: I think the original meaning of this part was that the transient lifetime (in general) does not satisfy the "should be reused within a component" requirement. Not sure what you mean by adding the "with disposable transient services" here?

The whole intro to `OwningComponentBase` here is that you can inject a composition root (`ScopedService` or by using `Services`) with lifetime aligned to the component. This is something neither the transient nor the singleton lifetime can offer (when we're talking about regular injection).

Uff, not an easy topic. :-D

Author: > Not sure what you mean by adding the "with disposable transient services" here?
>
> neither the transient nor the singleton lifetime can offer (when we're talking about regular injection).

That wasn't my interpretation of Tanay's remarks. BTW, I thought it was Steve that wrote the original design proposal, but it was Tanay, who has since moved to a different team. Anyway, I added "disposable transient services" here because according to Tanay it's ok to use non-disposable transient services. See the first column of his table ...

> Transient Service which do NOT implement IDisposable ... Are garbage collected when the component in which the service was injected is disposed.

https://github.com/dotnet/aspnetcore/issues/26676#issuecomment-1179447445
